akka-stream/src/main/scala/akka/stream/stage/Stage.scala

```diff
@@ -455,19 +514,64 @@ sealed trait Context[Out] {
  * events making the balance right again: 1 hold + 1 external event = 2 external event
  */
 trait DetachedContext[Out] extends Context[Out] {
-  def hold(): FreeDirective
+  def holdUpstream(): UpstreamDirective
+  def holdUpstreamAndPush(elem: Out): UpstreamDirective
+
+  def holdDownstream(): DownstreamDirective
+  def holdDownstreamAndPull(): DownstreamDirective
 
   /**
    * This returns `true` when [[#hold]] has been used
    * and it is reset to `false` after [[#pushAndPull]].
    */
-  def isHolding: Boolean
+  def isHoldingBoth: Boolean = isHoldingUpstream && isHoldingDownstream
+  def isHoldingUpstream: Boolean
+  def isHoldingDownstream: Boolean
 
   def pushAndPull(elem: Out): FreeDirective
 
 }
 
 /**
+ * An asynchronous callback holder that is attached to an [[AsyncContext]].
+ * Invoking [[AsyncCallback#invoke]] will eventually lead to [[AsyncStage#onAsyncInput]]
+ * being called.
+ */
+trait AsyncCallback[T] {
```

Comment:
Yes, Function1 has a ton of methods that this one does not need. Plus: Java API :-)